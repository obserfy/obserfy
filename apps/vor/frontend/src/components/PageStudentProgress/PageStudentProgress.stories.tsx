import { FC } from "react"
import PageStudentProgress from "./PageStudentProgress"

export default {
  title: "Page/PageStudentProgress",
  component: PageStudentProgress,
  parameters: {
    componentSubtitle: "Just a simple PageStudentProgress",
  },
}

export const Basic: FC = () => (
  <PageStudentProgress areaId="asdfasfd" studentId="asdfadsf" />
)
