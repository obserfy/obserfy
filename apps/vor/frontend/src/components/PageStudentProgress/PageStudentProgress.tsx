import { FC, useState } from "react"
import { Box, Card, Flex } from "theme-ui"
import { useGetStudent } from "../../hooks/api/useGetStudent"
import Typography from "../Typography/Typography"
import { BackNavigation } from "../BackNavigation/BackNavigation"
import { useGetArea } from "../../hooks/api/useGetArea"
import { Subject, useGetAreaSubjects } from "../../hooks/api/useGetAreaSubjects"
import {
  Material,
  useGetSubjectMaterials,
} from "../../hooks/api/useGetSubjectMaterials"

import Icon from "../Icon/Icon"
import { ReactComponent as NextIcon } from "../../icons/next-arrow.svg"
import {
  MaterialProgress,
  materialStageToString,
  useGetStudentAssessments,
} from "../../hooks/api/useGetStudentAssessments"
import Pill from "../Pill/Pill"
import StudentMaterialProgressDialog from "../StudentMaterialProgressDialog/StudentMaterialProgressDialog"
import LoadingPlaceholder from "../LoadingPlaceholder/LoadingPlaceholder"
import { STUDENT_OVERVIEW_URL, STUDENTS_URL } from "../../routes"
import BackButton from "../BackButton/BackButton"
import Breadcrumb from "../Breadcrumb/Breadcrumb"
import BreadcrumbItem from "../Breadcrumb/BreadcrumbItem"

interface Props {
  studentId: string
  areaId: string
}
export const PageStudentProgress: FC<Props> = ({ areaId, studentId }) => {
  const area = useGetArea(areaId)
  const student = useGetStudent(studentId)
  const subjects = useGetAreaSubjects(areaId)
  const [isEditing, setIsEditing] = useState(false)
  const [selectedMaterial, setSelectedMaterial] = useState<Material>()
  const progress = useGetStudentAssessments(studentId)
  const loading =
    student.status === "loading" ||
    area.status === "loading" ||
    subjects.status === "loading" ||
    progress.status === "loading"

  const backNavigation = (
    <BackNavigation
      text="Student Details"
      to={STUDENT_OVERVIEW_URL(studentId)}
    />
  )

  if (loading) {
    return (
      <Box sx={{ maxWidth: "maxWidth.sm" }} m={[3, "auto"]}>
        {backNavigation}
        <LoadingPlaceholder
          sx={{ width: "100%", height: "6rem" }}
          mb={2}
          mt={4}
        />
        <LoadingPlaceholder sx={{ width: "100%", height: "6rem" }} mb={4} />
        <LoadingPlaceholder sx={{ width: "100%", height: "6rem" }} mb={2} />
        <LoadingPlaceholder sx={{ width: "100%", height: "6rem" }} mb={2} />
        <LoadingPlaceholder sx={{ width: "100%", height: "6rem" }} mb={2} />
        <LoadingPlaceholder sx={{ width: "100%", height: "6rem" }} mb={2} />
      </Box>
    )
  }

  const selectedProgress = progress.data?.find(
    ({ materialId }) => materialId === selectedMaterial?.id
  )

  return (
    <>
      <Box sx={{ maxWidth: "maxWidth.sm" }} margin="auto" pb={5}>
        <Flex sx={{ height: 48, alignItems: "center" }}>
          <BackButton to={STUDENT_OVERVIEW_URL(studentId)} />
          <Breadcrumb>
            <BreadcrumbItem to={STUDENTS_URL}>Students</BreadcrumbItem>
            <BreadcrumbItem to={STUDENT_OVERVIEW_URL(studentId)}>
              {student.data?.name.split(" ")[0]}
            </BreadcrumbItem>
            <BreadcrumbItem>{` ${area.data?.name} Progress`}</BreadcrumbItem>
          </Breadcrumb>
        </Flex>
        {subjects.data?.map((subject) => (
          <Card
            mb={4}
            key={subject.id}
            mx={[0, 3]}
            sx={{ borderRadius: [0, "default"] }}
          >
            <Typography.H6 p={3}>{subject.name}</Typography.H6>
            <SubjectMaterials
              subject={subject}
              progress={progress.data ?? []}
              onMaterialClick={(material) => {
                setSelectedMaterial(material)
                setIsEditing(true)
              }}
            />
          </Card>
        ))}
      </Box>
      {isEditing && (
        <StudentMaterialProgressDialog
          studentId={studentId}
          stage={selectedProgress?.stage}
          lastUpdated={selectedProgress?.updatedAt}
          materialName={selectedMaterial?.name ?? ""}
          materialId={selectedMaterial?.id ?? ""}
          onDismiss={() => setIsEditing(false)}
        />
      )}
    </>
  )
}

const SubjectMaterials: FC<{
  subject: Subject
  progress: MaterialProgress[]
  onMaterialClick: (material: Material) => void
}> = ({ progress, subject, onMaterialClick }) => {
  const materials = useGetSubjectMaterials(subject.id)

  if (materials.status === "loading") {
    return (
      <>
        <LoadingPlaceholder sx={{ width: "100%", height: "6rem" }} mb={2} />
        <LoadingPlaceholder sx={{ width: "100%", height: "6rem" }} mb={2} />
        <LoadingPlaceholder sx={{ width: "100%", height: "6rem" }} mb={2} />
        <LoadingPlaceholder sx={{ width: "100%", height: "6rem" }} mb={2} />
        <LoadingPlaceholder sx={{ width: "100%", height: "6rem" }} mb={2} />
        <LoadingPlaceholder sx={{ width: "100%", height: "6rem" }} mb={2} />
      </>
    )
  }

  return (
    <>
      {materials.data?.map((material) => {
        const match = progress.find((item) => item.materialId === material.id)
        const stage = materialStageToString(match?.stage)

        return (
          <Flex
            key={material.id}
            py={2}
            onClick={() => onMaterialClick(material)}
            sx={{
              alignItems: "center",
              cursor: "pointer",
              borderTopColor: "border",
              borderTopWidth: "1px",
              borderTopStyle: "solid",
            }}
          >
            <Typography.Body ml={3} sx={{ fontSize: 1 }} mr="auto">
              {material.name}
            </Typography.Body>
            {stage && (
              <Pill
                mx={2}
                text={stage}
                color={`materialStage.on${stage}`}
                backgroundColor={`materialStage.${stage.toLocaleLowerCase()}`}
              />
            )}
            <Icon as={NextIcon} mr={3} />
          </Flex>
        )
      })}
    </>
  )
}

export default PageStudentProgress
