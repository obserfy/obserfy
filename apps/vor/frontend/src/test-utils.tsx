/* eslint-disable import/no-extraneous-dependencies */
import { FC, ReactElement } from "react"
import { QueryClient, QueryClientProvider } from "react-query"
import { ThemeProvider, Box, Theme } from "theme-ui"
import { render, RenderOptions, RenderResult } from "@testing-library/react"
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router"
import { I18nProvider } from "@lingui/react"
import { i18n } from "@lingui/core"
import { en } from "make-plural/plurals"
import theme from "./gatsby-plugin-theme-ui"
// @ts-ignore
import enCatalog from "../i18n/lingui/en/messages.js"

// any used here because the catalogs are generated b lingui
// not our responsibility
i18n.loadLocaleData("en", { plurals: en })
i18n.load("en", enCatalog.messages as any)
i18n.activate("en")

const queryClient = new QueryClient()
const history = createHistory(createMemorySource("/"))

const AllProviders: FC = ({ children }) => {
  return (
    <I18nProvider i18n={i18n}>
      <QueryClientProvider client={queryClient}>
        <LocationProvider history={history}>
          <ThemeProvider theme={theme as Theme}>
            <Box sx={{ fontSize: [16, 20] }}>{children}</Box>
          </ThemeProvider>
        </LocationProvider>
      </QueryClientProvider>
    </I18nProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: AllProviders, ...options })

// re-export everything
export * from "@testing-library/react"

// override render method
export { customRender as render }
