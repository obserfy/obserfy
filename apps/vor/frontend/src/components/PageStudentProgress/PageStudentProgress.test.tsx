import { render } from "../../test-utils"
import { Basic } from "./PageStudentProgress.stories"

describe("PageStudentProgress", () => {
  it("should render correctly", () => {
    // TODO: This is a temporary fix, make sure to test mock response correctly.
    const response: string[] = []
    fetchMock.mockResponse(JSON.stringify(response))
    const { container } = render(<Basic />)
    expect(container).toMatchSnapshot()
  })
})
