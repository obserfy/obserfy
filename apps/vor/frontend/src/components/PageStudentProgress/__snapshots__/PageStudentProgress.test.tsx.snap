// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageStudentProgress should render correctly 1`] = `
@keyframes animation-0 {
  0% {
    opacity: 0.05;
  }

  50% {
    opacity: 0.2;
  }

  100% {
    opacity: 0.05;
  }
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 16px;
}

@media screen and (min-width: 40em) {
  .emotion-0 {
    font-size: 20px;
  }
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin: 16px;
  max-width: 640px;
}

@media screen and (min-width: 40em) {
  .emotion-1 {
    margin: auto;
  }
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-left: -8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  margin: 16px;
  margin-right: 4px;
  fill: var(--theme-ui-colors-icon);
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  fill: var(--theme-ui-colors-textMediumEmphasis);
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: var(--theme-ui-colors-text);
  font-size: 14px;
  text-rendering: optimizeLegibility;
  line-height: 1.5;
  margin-bottom: 0;
  color: var(--theme-ui-colors-textMediumEmphasis);
}

.emotion-5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-bottom: 8px;
  margin-top: 32px;
  border-radius: 6px;
  -webkit-animation: 4s infinite animation-0;
  animation: 4s infinite animation-0;
  background-color: #a5a5a5;
  width: 100%;
  height: 6rem;
}

.emotion-6 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-bottom: 32px;
  border-radius: 6px;
  -webkit-animation: 4s infinite animation-0;
  animation: 4s infinite animation-0;
  background-color: #a5a5a5;
  width: 100%;
  height: 6rem;
}

.emotion-7 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-bottom: 8px;
  border-radius: 6px;
  -webkit-animation: 4s infinite animation-0;
  animation: 4s infinite animation-0;
  background-color: #a5a5a5;
  width: 100%;
  height: 6rem;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <a
        href="/dashboard/students/overview?studentId=asdfadsf"
        style="display: block;"
      >
        <div
          class="emotion-2"
        >
          <div
            class="emotion-3"
          />
          <p
            class="emotion-4"
          >
            Student Details
          </p>
        </div>
      </a>
      <div
        class="emotion-5"
      />
      <div
        class="emotion-6"
      />
      <div
        class="emotion-7"
      />
      <div
        class="emotion-7"
      />
      <div
        class="emotion-7"
      />
      <div
        class="emotion-7"
      />
    </div>
  </div>
</div>
`;
